planner/core/memtable_predicate_extractor.go

```diff
@@ -205,10 +238,159 @@ type ClusterConfigTableExtractor struct {
+type ClusterLogTableExtractor struct {
+	extractHelper
+
+	// SkipRequest means the where clause always false, we don't need to request any component
+	SkipRequest bool
+
+	// NodeTypes represents all components types we should send request to.
+	// e.g:
+	// 1. SELECT * FROM cluster_log WHERE type='tikv'
+	// 2. SELECT * FROM cluster_log WHERE type in ('tikv', 'tidb')
+	NodeTypes set.StringSet
+
+	// Addresses represents all components addresses we should send request to.
+	// e.g:
+	// 1. SELECT * FROM cluster_log WHERE address='192.168.1.7:2379'
+	// 2. SELECT * FROM cluster_log WHERE type in ('192.168.1.7:2379', '192.168.1.9:2379')
+	Addresses set.StringSet
+
+	// StartTime represents the beginning time of log message
+	// e.g: SELECT * FROM cluster_log WHERE time>'2019-10-10 10:10:10.999'
+	StartTime int64
+	// EndTime represents the ending time of log message
+	// e.g: SELECT * FROM cluster_log WHERE time<'2019-10-11 10:10:10.999'
+	EndTime int64
+	// Pattern is used to filter the log message
+	// e.g:
+	// 1. SELECT * FROM cluster_log WHERE message like '%gc%'
+	// 2. SELECT * FROM cluster_log WHERE message regexp '.*'
+	Patterns  []string
+	LogLevels set.StringSet
+}
+
+// Extract implements the MemTablePredicateExtractor Extract interface
+func (e *ClusterLogTableExtractor) Extract(
+	ctx sessionctx.Context,
+	schema *expression.Schema,
+	names []*types.FieldName,
+	predicates []expression.Expression,
+) []expression.Expression {
+	// Extract the `type/address` columns
+	remained, typeSkipRequest, nodeTypes := e.extractCol(schema, names, predicates, "type", true)
+	remained, addrSkipRequest, addresses := e.extractCol(schema, names, remained, "address", false)
+	remained, levlSkipRequest, logLevels := e.extractCol(schema, names, remained, "level", false)
+	e.SkipRequest = typeSkipRequest || addrSkipRequest || levlSkipRequest
+	e.NodeTypes = nodeTypes
+	e.Addresses = addresses
+	e.LogLevels = logLevels
+	if e.SkipRequest {
+		return nil
+	}
+
+	// Extract the `time/message/level` columns
+	const (
+		ColNameTime    = "time"
+		ColNameMessage = "message"
+	)
+	extractCols := make(map[int64]*types.FieldName)
+	for i, name := range names {
+		if ln := name.ColName.L; ln == ColNameTime || ln == ColNameMessage {
+			extractCols[schema.Columns[i].UniqueID] = name
+		}
+	}
+
+	// We should compile the regexp to (?=expr1)(?=expr2) if there are more than one regular expressions
+	// e.g:
+	// SELECT * FROM cluster_log WHERE message LIKE '%a%' AND message LIKE '%b%' AND message REGEXP 'gc.*[0-9]{10,20}'
+	var patterns []string
+	var startTime, endTime int64
+	for _, expr := range predicates {
+		x, ok := expr.(*expression.ScalarFunction)
+		if !ok {
+			remained = append(remained, expr)
+			continue
+		}
+		var colName string
+		var datums []types.Datum
+
+		switch x.FuncName.L {
+		case ast.GT, ast.GE, ast.LT, ast.LE, ast.EQ, ast.Like, ast.Regexp:
+			colName, datums = e.extractColBinaryOpConsExpr(extractCols, x)
+		}
+
+		switch colName {
```

Comment:
Extracting pattern and time is complicated, and they're probably used in other extractors.
For example, you can define a function like 
`func extractStringPattern(x *expression.ScalarFunction, stringDatum Datum) (remained []expression.Expression, patterns []string)`
`func extractTime(x *expression.ScalarFunction, timeDatum Datum) (remained []expression.Expression, startTime int64, endTime int64)`

Reply:
Good idea.

Reply:
done